### PR TITLE
test(nemesis): added FreeTierSetMonkey

### PR DESCRIPTION
**Pull request by soyacz**
For free tier cloud we want to limit disruptions to ones that may actually happen. So we want to skip disruptions for mgmt, topology changes (except ones that may happen due failures) or any other that are unlikely to happen on Scylla Cloud's free tier.

New Monkey `FreeTierSetMonkey` has been created which behaves similar to SisyphusMonkey but on limited set. Disruptions eliglible for free tier are marked with new flag `free_tier_set` in `Nemesis` class so when new nemesis are added it's easier to remember to update `FreeTierSetMonkey`.

refs: https://github.com/scylladb/qa-tasks/issues/910

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by soyacz** [comment]
Nemesis that are skipped from the list:
```
MgmtBackup - don't use mgmt
MgmtBackupSpecificKeyspaces - don't use mgmt
MgmtRepair - don't use mgmt
NodeTerminateAndReplace - unsupported on K8s
OperatorNodeTerminateAndReplace - exclusive
OperatorNodeTerminateDecommissionAdd - exclusive
NodeRestartWithResharding - free tier should not have resharding needed
RollingRestartConfigChangeInternodeCompression - unsupported on K8s
SwitchBetweenPasswordAuthAndSaslauthdAuth - not used on Free tier (auth using bundle)
RandomInterruptionNetworkMonkey - no extra network interface
RejectNodeExporterNetworkMonkey - unsupported on K8s
RejectThriftNetworkMonkey - unsupported on K8s
StopStartInterfacesNetworkMonkey - no extra network interface
BlockNetworkMonkey - no extra network interface
DecommissionStreamingErrMonkey - unsupported on K8s
StartStopMajorCompaction - my opinion
StartStopScrubCompaction - my opinion
StartStopCleanupCompaction - my opinion
StartStopValidationCompaction - my opinion
DecommissionMonkey - not used on Free tier
DecommissionSeedNode - not used on Free tier
LoadAndStreamMonkey - not used on Free tier
RemoveServiceLevelMonkey - only for SLA
EnospcMonkey - it is disabled
EnospcAllNodesMonkey  - it is disabled
TerminateAndRemoveNodeMonkey - not used on Free tier
```

For ones that fit to new monkey, see the code changes.

**Comment by fruch** [comment]
> Nemesis that are skipped from the list:
> ```
> MgmtBackup - don't use mgmt
> MgmtBackupSpecificKeyspaces - don't use mgmt
> MgmtRepair - don't use mgmt
> NodeTerminateAndReplace - unsupported on K8s
> OperatorNodeTerminateAndReplace - exclusive
> OperatorNodeTerminateDecommissionAdd - exclusive
> NodeRestartWithResharding - free tier should not have resharding needed
> RollingRestartConfigChangeInternodeCompression - unsupported on K8s
> SwitchBetweenPasswordAuthAndSaslauthdAuth - not used on Free tier (auth using bundle)
> RandomInterruptionNetworkMonkey - no extra network interface
> RejectNodeExporterNetworkMonkey - unsupported on K8s
> RejectThriftNetworkMonkey - unsupported on K8s
> StopStartInterfacesNetworkMonkey - no extra network interface
> BlockNetworkMonkey - no extra network interface
> DecommissionStreamingErrMonkey - unsupported on K8s
> StartStopMajorCompaction - my opinion
> StartStopScrubCompaction - my opinion
> StartStopCleanupCompaction - my opinion
> StartStopValidationCompaction - my opinion
> DecommissionMonkey - not used on Free tier
> DecommissionSeedNode - not used on Free tier
> LoadAndStreamMonkey - not used on Free tier
> RemoveServiceLevelMonkey - only for SLA
> EnospcMonkey - it is disabled
> EnospcAllNodesMonkey  - it is disabled
> TerminateAndRemoveNodeMonkey - not used on Free tier
> ```
> 
> For ones that fit to new monkey, see the code changes.

@soyacz can you attached the list of what is selected to the test plan ? @roydahan was asking for it

**Comment by soyacz** [comment]
list of selected disruptions:
```
disrupt_replace_node_kubernetes
disrupt_add_drop_column
disrupt_repair_streaming_err
disrupt_delete_overlapping_row_ranges
disrupt_resetlocalschema
disrupt_toggle_table_ics
disrupt_toggle_table_gc_mode
disrupt_memory_stress
disrupt_run_cdcstressor_tool
disrupt_major_compaction
disrupt_show_toppartitions
disrupt_nodetool_cleanup
disrupt_modify_table
disrupt_corrupt_then_scrub
disrupt_truncate
disrupt_toggle_cdc_feature_properties_on_table
disrupt_abort_repair
disrupt_no_corrupt_repair
disrupt_kill_scylla
disrupt_rolling_restart_cluster
disrupt_nodetool_refresh
disrupt_snapshot_operations
disrupt_nodetool_flush_and_reshard_on_kubernetes
disrupt_truncate_large_partition
disrupt_rebuild_streaming_err
disrupt_delete_by_rows_range
```

File: sdcm/nemesis.py
Comment (fruch) [reply sdcm/nemesis.py:4977]: and fix `build_list_of_disruptions_to_execute` to actually use the `nemesis_selector` parameter, and then you could do:

```
class FreeTierSetMonkey(SisyphusMonkey):
     def __init__(self, *args, *kwargs):
             super(SisyphusMonkey, self).__init__(*args, **kwargs) # skip SisyphusMonkey __init__ 
            self.build_list_of_disruptions_to_execute(nemesis_selector=['free_tier_set', '!topology_changes'])
            self.shuffle_list_of_disruptions()
```

Comment (fgelcer) [reply sdcm/nemesis.py:4977]: we could use `SisyphusMonkey` itself, with filters, no?

Comment (fruch) [reply sdcm/nemesis.py:4977]: > we could use `SisyphusMonkey` itself, with filters, no?

technically yes, but this way it has a definition and a name, and configuration would be clearer

```yaml
nemesis_class: ['FreeTierSetMonkey', 'FreeTierSetMonkey']
```
vs.
```yaml
nemesis_class: ['SisyphusMonkey', 'SisyphusMonkey']
nemesis_selector: [ ['free_tier_set', '!topology_changes'], ['free_tier_set', '!topology_changes'] ]
````

or

```yaml
nemesis_class: ['SisyphusMonkey', 'NoOpMonkey']
nemesis_selector: [ ['free_tier_set', '!topology_changes'], [] ]
````

**Comment by fruch** [comment]
> disrupt_replace_node_kubernetes
> disrupt_add_drop_column
> disrupt_repair_streaming_err

repair is same as manager repair from my POV, it human operator action, not user action

> disrupt_delete_overlapping_row_ranges
> disrupt_resetlocalschema
> disrupt_toggle_table_ics
> disrupt_toggle_table_gc_mode
> disrupt_memory_stress
> disrupt_run_cdcstressor_tool
> disrupt_major_compaction

major is done via nodetool it human operator action, not user action

> disrupt_show_toppartitions
> disrupt_nodetool_cleanup

cleanup is done via nodetool its human operator action, not user action

> disrupt_modify_table
> disrupt_corrupt_then_scrub
> disrupt_truncate
> disrupt_toggle_cdc_feature_properties_on_table
> disrupt_abort_repair
> disrupt_no_corrupt_repair

repair is done via nodetool, its human operator action, not user action

> disrupt_kill_scylla
> disrupt_rolling_restart_cluster
> disrupt_nodetool_refresh

refresh is done via nodetool, its human operator action, not user action

> disrupt_snapshot_operations

snapshot is done via nodetool, its human operator action, not user action

> disrupt_nodetool_flush_and_reshard_on_kubernetes

flush is done via nodetool, its human operator action, not user action, and resharding isn't relevant (we gonna have 1 CPU per pod, and we won't be scaling that up) 

> disrupt_truncate_large_partition
> disrupt_rebuild_streaming_err

rebuild is done via nodetool, its human operator action, not user action

> disrupt_delete_by_rows_range

**Comment by fruch** [comment]
those are the list of candidates for implementing equivalents with chaos.mesh:

disrupt_memory_stress - is current broken for lastest scylla docker images (no snapd)
RandomInterruptionNetworkMonkey - no extra network interface
RejectNodeExporterNetworkMonkey - unsupported on K8s
RejectThriftNetworkMonkey - unsupported on K8s
StopStartInterfacesNetworkMonkey - no extra network interface
BlockNetworkMonkey - no extra network interface
EnospcMonkey - it is disabled (cause of a scylla bug)
EnospcAllNodesMonkey  - it is disabled

**Comment by roydahan** [comment]
Please wait for my review as well

**Comment by soyacz** [comment]
I've updated PR according to Israel's suggestions. New list:
```
disrupt_toggle_table_gc_mode
disrupt_add_drop_column
disrupt_toggle_cdc_feature_properties_on_table
disrupt_delete_by_rows_range
disrupt_modify_table
disrupt_memory_stress
disrupt_corrupt_then_scrub
disrupt_delete_10_full_partitions
disrupt_rolling_restart_cluster
disrupt_resetlocalschema
disrupt_run_cdcstressor_tool
disrupt_toggle_table_ics
disrupt_abort_repair
disrupt_replace_node_kubernetes
disrupt_kill_scylla
disrupt_truncate_large_partition
disrupt_show_toppartitions
disrupt_truncate
disrupt_delete_overlapping_row_ranges
```

**Comment by soyacz** [comment]
And list of skipped disrupitons (for review):
```
disrupt_nodetool_flush_and_reshard_on_kubernetes
disrupt_start_stop_validation_compaction
disrupt_start_stop_major_compaction
disrupt_network_random_interruptions
disrupt_stop_start_scylla_server
disrupt_major_compaction
disrupt_decommission_streaming_err
disrupt_start_stop_cleanup_compaction
disrupt_nodetool_refresh
disrupt_rebuild_streaming_err
disrupt_switch_between_password_authenticator_and_saslauthd_authenticator_and_back
disrupt_destroy_data_then_rebuild
disrupt_nodetool_seed_decommission
disrupt_network_block
disrupt_add_remove_dc
disrupt_snapshot_operations
disrupt_repair_streaming_err
disrupt_nodetool_decommission
disrupt_ldap_connection_toggle
disrupt_mgmt_backup
disrupt_run_unique_sequence
disrupt_remove_service_level_while_load
disrupt_network_reject_thrift
disrupt_nodetool_drain
disrupt_network_reject_node_exporter
disrupt_disable_enable_ldap_authorization
disrupt_remove_node_then_add_node
disrupt_no_corrupt_repair
disrupt_destroy_data_then_repair
disrupt_restart_with_resharding
disrupt_load_and_stream
disrupt_grow_shrink_new_rack
disrupt_start_stop_scrub_compaction
disrupt_nodetool_cleanup
disrupt_terminate_and_replace_node
disrupt_mgmt_repair_cli
disrupt_hot_reloading_internode_certificate
disrupt_grow_shrink_cluster
disrupt_rolling_config_change_internode_compression
disrupt_stop_wait_start_scylla_server
disrupt_network_start_stop_interface
disrupt_mgmt_backup_specific_keyspaces
disrupt_terminate_and_replace_node_kubernetes
disrupt_terminate_decommission_add_node_kubernetes
disrupt_nodetool_enospc
```

File: sdcm/nemesis.py
Comment (vponomaryov) [inline sdcm/nemesis.py:4978]: Would be good to see "why" info for skipping that.

File: sdcm/nemesis.py
Comment (vponomaryov) [inline sdcm/nemesis.py:4697]: This nemesis is absent in the list you posted here: https://github.com/scylladb/scylla-cluster-tests/pull/5504#issuecomment-1337171769
Which one is correct?

File: sdcm/nemesis.py
Comment (vponomaryov) [inline sdcm/nemesis.py:4609]: This nemesis is absent in the list you posted here: https://github.com/scylladb/scylla-cluster-tests/pull/5504#issuecomment-1337171769
Which one is correct?

Comment (soyacz) [reply sdcm/nemesis.py:4609]: I created list automatically as it is seen by SCT - these nemesis are absent because they are disabled due issues. 
Disruption method names are not starting with `self.disrupt_` so SCT does not run them in random monkey tests.

**Comment by fruch** [comment]
@roydahan ping, you wanted to review this list before we merge.

**Comment by vponomaryov** [comment]
> @soyacz let run this nemesis on the multiteant test case we have, to see it's operating as we intended

I ran it with 14 clients in 12h job here -> https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/valerii/job/vp-longevity-scylla-operator-12h-multitenant-eks/19/consoleFull
It failed, I haven't investigated it yet.

**Comment by soyacz** [comment]
Looks it is operating as intended, you can see nemesis list in Argus: https://argus.scylladb.com/test/ae47aebc-b864-4e31-a2d7-407351dc9544/runs?additionalRuns[]=bc49d6de-d474-4622-8d39-5082725587f1

Test survived 10 hours of Nemesis - critically failed due `Operation x10 on key(s)` c-s on missing write quorum errors. Also during the run we can see some errors related to cluster overload (too many inflight hints) and truncate table timeout.

File: test-cases/scylla-operator/longevity-scylla-operator-3h-multitenant.yaml
Comment (roydahan) [inline test-cases/scylla-operator/longevity-scylla-operator-3h-multitenant.yaml:27]: Shouldn't it be SisyphusMonket with a nemesis_selector: 'free_tier_set'?

**Comment by roydahan** [comment]
> I've updated PR according to Israel's suggestions. New list:
> 
> ```
> disrupt_add_drop_column
> disrupt_run_cdcstressor_tool
> disrupt_delete_by_rows_range
> disrupt_soft_reboot_node
> disrupt_toggle_table_ics
> disrupt_toggle_table_gc_mode
> disrupt_delete_10_full_partitions
> disrupt_toggle_cdc_feature_properties_on_table
> disrupt_rolling_restart_cluster
> disrupt_multiple_hard_reboot_node
> disrupt_memory_stress
> disrupt_replace_node_kubernetes
> disrupt_truncate
> disrupt_resetlocalschema
> disrupt_delete_overlapping_row_ranges
> disrupt_modify_table
> disrupt_hard_reboot_node
> disrupt_kill_scylla
> disrupt_truncate_large_partition
> ```


The list looks good to me, although it's mixing two types of nemesis.
1. The type that is done from the user side and we can't control.
2. reboots / replace k8s, which may be out of scope for the freeTier.

Maybe we want to seprate to "free_tier_set", "Disruptive" and "free_tier_set", "Non-Disruptive"?
I'm not sure...

**Review by vponomaryov — APPROVED** [review]

**Comment by roydahan** [comment]
Ok.

**Review by roydahan — APPROVED** [review]

File: sdcm/nemesis.py
Comment (roydahan) [inline sdcm/nemesis.py:4969]: @soyacz / @fruch, 
While working on nemesis.py I realized I don't understand why we need this MonkeySet.

Why don't we use SisyphusMonkey with nemesis_selector in the original test yaml?

Comment (fruch) [reply sdcm/nemesis.py:4969]: See
https://github.com/scylladb/scylla-cluster-tests/pull/5504#discussion_r1035603657

for the reason, it simply the configuration of multi-tenant setups.

Comment (roydahan) [reply sdcm/nemesis.py:4969]: thx

Comment (vponomaryov) [reply sdcm/nemesis.py:4969]: Looks like worth a comment in the code for that new nemesis.

**Pull request merged**
